### PR TITLE
Add Guided Jacobi Iterative Solver

### PR DESCRIPTION
**Pull request by IgorOchocki**
# Adding a New Sample(s)

## Description

The sample is a guided Jacobi Iterative solver showcasing the benefits of a GPU and a multi GPU offload

## Checklist
Administrative
- [ ] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): <insert Name Here>
- [ ] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager

Code Development
- [ ] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [ ] Adhere to readme template 
- [ ] Enforce format via clang-format config file
- [ ] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [ ] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [ ] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

**Comment by jimmytwei** [comment]
@tomlenth You reviewed the README with @IgorOchocki right? If so, mark it as approved.

**Review by tomlenth — DISMISSED** [review]
yes, worked extensively with John and Igor.  Great sample.

**Comment by jimmytwei** [comment]
@pmpeter1 Can you review this new guided Jacobi iterative solver sample? Thanks.

**Review by pmpeter1 — CHANGES_REQUESTED** [review]

**Comment by IgorOchocki** [comment]
Updated the sample with clang formatting.

File: DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/1_guided_jacobi_iterative_solver_cpu.cpp [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+#include <algorithm>
+#include <chrono>
+#include <cstdlib>
+#include <ctime>
+#include <oneapi/dpl/random>
+#include <vector>
+
+using namespace sycl;
+
+typedef double real;
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/1_guided_jacobi_iterative_solver_cpu.cpp:19]: should be real -> Real (it's a type).

File: DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/1_guided_jacobi_iterative_solver_cpu.cpp [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+// Function responsible for generating a float type
+// diagonally dominant matrix. Float had to be used
+// as using double would result in segmentation faults
+// for extreamlly large matrixes. This is also an example
+// of using sycl based RNG which had to be used as using
+// external (non sycl) functions slows down the execution
+// drasticly.
+void GenerateMatrix(std::vector<float> &input_matrix,
+                    std::vector<real> &input_results) {
+  for (int i = 0; i < kN; ++i) {
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/1_guided_jacobi_iterative_solver_cpu.cpp:41]: Can you use a more descriptibe name rather then "N" (kN).    This should be what does this represent.   N is a little too "Fortran-like" shorthand.

File: DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/2_guided_jacobi_iterative_solver_gpu.cpp [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+#include <algorithm>
+#include <chrono>
+#include <cstdlib>
+#include <ctime>
+#include <oneapi/dpl/random>
+#include <vector>
+
+using namespace sycl;
+
+typedef double real;
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/2_guided_jacobi_iterative_solver_gpu.cpp:19]: Should be real -> Real (its a type).

File: DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/2_guided_jacobi_iterative_solver_gpu.cpp [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+// Program variables, feel free to change anything .
+static const int kN = 30000;
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/2_guided_jacobi_iterative_solver_gpu.cpp:22]: Would be better to use a more descriptive name (not kN).

Comment (pmpeter1) [reply DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/2_guided_jacobi_iterative_solver_gpu.cpp:22]: The other named constants are much better in being descriptive.

File: DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+#include <algorithm>
+#include <chrono>
+#include <cstdlib>
+#include <ctime>
+#include <oneapi/dpl/random>
+#include <vector>
+
+using namespace sycl;
+
+typedef double real;
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp:19]: Same changes for "real" and "kN".

File: DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+std::ofstream outfile;
+
+// Function responsible for generating a float type
+// diagonally dominant matrix. Float had to be used
+// as using double would result in segmentation faults
+// for extreamlly large matrixes. This is also an example
+// of using sycl based RNG which had to be used as using
+// external (non sycl) functions slows down the execution
+// drasticly.
+void GenerateMatrix(std::vector<float> &input_matrix,
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp:39]: Can the input matrix be a const refernce?

File: DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+        }
+      }
+    }
+  }
+
+  buffer bufin_mat(input_matrix);
+  buffer bufin_res(input_results);
+
+  q[0].submit([&](handler &h) {
+    accessor M{bufin_mat, h};
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp:60]: Should be lower case or snake_case (nor M, R).   Also not descriptive enough.

File: DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+      else
+        M[i * kN + i] = -1 * (sum + 1);
+
+      R[i] = distr(engine);
+      R[i] = round(100. * R[i]) / 100.;
+    });
+  });
+}
+// Function responsible for printing the matrix, called only for N < 10.
+void PrintMatrix(std::vector<float> input_matrix,
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp:125]: Should input_matrix be const reference?

File: DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+  for (int i = 0; i < kN; ++i) {
+    outfile << '[';
+    for (int j = i * kN; j < kN * (i + 1); ++j) {
+      outfile << input_matrix[j] << " ";
+    }
+    outfile << "][" << input_results[i] << "]\n";
+  }
+}
+// Function responsible for printing the results.
+void PrintResults(std::vector<real> data, int N) {
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp:144]: Should data be const reference?

File: DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+  // The main functionality of the Jacobi Solver. Every iteration
+  // calculates new values until the difference between the values
+  // calculatedthis iteration and the one before is less than the error.
+  {
+    do {
+      buffer bufout_data(output_data);
+      buffer bufold_out_data(old_output_data);
+      for (int i = 0; i < kN; ++i) old_output_data[i] = output_data[i];
+      q[0].submit([&](handler &h) {
+            accessor D{bufout_data, h};
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp:231]: D, OV, M, R should follow variable name conventions and be a little more descriptive.

File: DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+                if (z != i) D[i] = D[i] - (OV[z] * static_cast<real>(M[j]));
+                j = j + 1;
+              }
+              D[i] = D[i] / static_cast<real>(M[it]);
+            });
+          })
+          .wait();
+
+      q[1].submit([&](handler &h) {
+            accessor D{bufout_data, h};
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/DenseLinearAlgebra/guided_jacobi_iterative_gpu_optimization/src/3_guided_jacobi_iterative_solver_multi_gpu.cpp:251]: Same for D. OV, M, R.     This are not idiomatic like "id" or "i", "j", "k" (loop counters).

**Review by pmpeter1 — CHANGES_REQUESTED** [review]

**Review by pmpeter1 — APPROVED** [review]

**Pull request merged**